src/macaron/slsa_analyzer/ci_service/github_actions/github_actions_ci.py

```diff
@@ -624,3 +462,120 @@ def has_kws_in_log(self, latest_run: dict, build_log: list) -> bool:
+            The path to the repo.
+        macaron_path: str
+            Macaron's root path (optional).
+
+        Returns
+        -------
+        CallGraph: CallGraph
+            The call graph built for GitHub Actions.
+        """
+        if not macaron_path:
+            macaron_path = global_config.macaron_path
+
+        root: BaseNode = BaseNode()
+        gh_cg = CallGraph(root, repo_path)
+
+        # Parse GitHub Actions workflows.
+        files = self.get_workflows(repo_path)
+        for workflow_path in files:
+            try:
+                callee = build_call_graph_from_path(
+                    root=root, workflow_path=workflow_path, repo_path=repo_path, macaron_path=macaron_path
+                )
+            except ParseError:
+                logger.debug("Skip adding workflow at %s to the callgraph.", workflow_path)
+                continue
+            root.add_callee(callee)
+        return gh_cg
+
+    def get_build_tool_commands(self, callgraph: CallGraph, build_tool: BaseBuildTool) -> Iterable[BuildToolCommand]:
+        """
+        Traverse the callgraph and find all the reachable build tool commands.
+
+        Parameters
+        ----------
+        callgraph: CallGraph
+            The callgraph reachable from the CI workflows.
+        build_tool: BaseBuildTool
+            The corresponding build tool for which shell commands need to be detected.
+
+        Yields
+        ------
+        BuildToolCommand
+            The object that contains the build command as well useful contextual information.
+
+        Raises
+        ------
+        CallGraphError
+            Error raised when an error occurs while traversing the callgraph.
+        """
+        for node in callgraph.bfs():
+            # We are just interested in nodes that have bash commands.
+            if isinstance(node, BashNode):
+                # We collect useful contextual information for the called BashNode.
+                caller_node = node.caller
+                # The GitHub Actions workflow that triggers the path in the callgraph.
+                workflow_node = None
+                # The GitHub Actions job that triggers the path in the callgraph.
+                job_node = None
+                # The step in GitHub Actions job that triggers the path in the callgraph.
+                step_node = node if node.node_type == BashScriptType.INLINE else None
+
+                # Walk up the grallgraph to find the relevant caller nodes.
+                while caller_node is not None:
+                    match caller_node:
+                        case GitHubWorkflowNode():
+                            workflow_node = caller_node
+                        case GitHubJobNode():
+                            job_node = caller_node
+                        case BashNode(node_type=BashScriptType.INLINE):
+                            step_node = caller_node
+
+                    caller_node = caller_node.caller
```

Comment:
I am not sure I get the intention behind this while block.
From my understanding, we are trying to obtain all nodes that are "transitive parents" of a BashNode in the call graph so that we can extract necessary information about the "path" that lead to a build tool command being invoked.
For example:
```yaml
# build.yaml
...

jobs:
  build:
     steps:
       -  run: ./build.sh
```
This will be represented in the call graph similar to this:
```
GitHubWorkflowNode(source_path="build.yaml") -> GitHubJobNode(name="build") -> BashNode(type=INLINE) -> (BashNode(type=FILE)) * <recursive_depth>. 
```
The reason we need to have this loop is because for `BashNode(type=FILE)`, we need to go up `<recursive_depth>` times before we can reach the `GitHub*` nodes right?
Even thought the current implementation is fine by itself, I have some suggestions:
- I think it would be a bit clearer if we also implementing the default case handling for the match case block
```python
                    match caller_node:
                        case GitHubWorkflowNode():
                            workflow_node = caller_node
                        case GitHubJobNode():
                            job_node = caller_node
                        case BashNode(node_type=BashScriptType.INLINE):
                            step_node = caller_node
                        case _:
                            # We only care about the GitHub* nodes and ignore other ones. 
                            pass
```
- It's also good to note down (as a comment) here that we are working on the assumption that `GitHubWorkflowNode`, `GitHubJobNode` and `BashNode(type=INLINE)` will appear once as transitive predecessor of any BashNode within the Call Graph.
- I think the assignment on this line https://github.com/oracle/macaron/blob/ed96daea0f0c4ae81f6f42b464d859f0829563aa/src/macaron/slsa_analyzer/ci_service/github_actions/github_actions_ci.py#L533 is duplicated with 
https://github.com/oracle/macaron/blob/ed96daea0f0c4ae81f6f42b464d859f0829563aa/src/macaron/slsa_analyzer/ci_service/github_actions/github_actions_ci.py#L542-L543

Reply:
Adding the default case as suggested doesn't make a difference, so we shouldn't add add it.

We don't make any assumption about the parent-child relationships in GitHub Actions. This callgraph is capturing the control flow based on how GitHub Actions works. I can add a comment to explain that.

The assignment is not duplicated. Think of a case where a `BashNode` is of type `BashScriptType.FILE`. The first assignment will set `step_node` to `None`. Then we walk up the callgraph and find an inlined `BashNode` caller node, which sets the `step_node`.

Reply:
> We don't make any assumption about the parent-child relationships in GitHub Actions. 

Then would it be possible if there is two `GitHubWorkflowNode` for example on the path when we walk up the call graph? In this case `workflow_node` will be overridden with the `highest` `GitHubWorkflowNode`. Is this behavior expected too?

Reply:
>  I can add a comment to explain that.

Adding a comment sounds good to me.

Reply:
> The assignment is not duplicated. Think of a case where a BashNode is of type BashScriptType.FILE. The first assignment will set step_node to None. Then we walk up the callgraph and find an inlined BashNode caller node, which sets the step_node.

Ah I see, Thanks for the clarification!

Reply:
> r we can safely assume that this scenario will never happen

This won't happen today, but in the future if we support analysis of external workflows, this could happen and the nodes in the main workflow will be capture. But whether this is suitable or not should be revisited later.

Reply:
Ah I see, then it would be good if we note that this scenario will not happen as of now, but it can change in the future as a comment in the code. But I don't have a strong opinion so it's fine as it is.

Reply:
See commit [6832dcd](https://github.com/oracle/macaron/pull/673/commits/6832dcd2d8a32a57558360b6c090ebdd8e7039a8)